docs/proposal_pytorch_distribution.md

```diff
@@ -0,0 +1,335 @@
+   PyTorch releases requires version bumps on existing packages - more detail
+   in strategy will be needed here.
+2. Tests for a package should be _runnable in a standardized way_, via
+   `conda-build --test`. This is easy to achieve via either a `test:` section
+   in the recipe (`meta.yaml`) or a `run_test.py` file. See [this section of
+   the conda-build docs](https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#test-section)
+   for details. An advantage of this method is that `conda-build` is already
+   aware of channels and dependencies, so it should work with very little
+   extra effort.
+
+
+### What happens when a new PyTorch release is made?
+
+For minor or major versions of PyTorch, new releases of downstream packages
+will also be necessary. A number of packages, such as `torchvision`,
+`torchaudio` and `torchtext`, are anyway released in sync. Other packages in
+the `pytorch` channel may need to be manually released via a PR to the
+`conda-distro` repo).
+
+Version constraints should be set such that a bugfix release of PyTorch does
+not require any new downstream package releases.
+
+
+### Dealing with packages that aren't maintained
+
+Proposing a package for inclusion in the `pytorch` channel implies a
+commitment to keep maintaining the package. There wil be a place to list one
+or more maintainers for each package so they can be pinged if needed. In case
+a package is not up-to-date or broken and it does not get fixed, after a
+certain duration (length TBD) it may be removed from the channel.
+
+
+## Alternatives
+
+### Conda-forge
+
+The main alternative to making the `pytorch` channel an integration channel
+that distributes many packages that depend on PyTorch is to have a
+(GPU-enabled) PyTorch package in conda-forge, and tell users and package
+authors that that is the place to go. It will require working with
+conda-forge in order to ensure that the `pytorch` package is of high quality,
+either by copying over the binaries from the `pytorch` channel or by
+migrating recipes and keeping them in sync. See
+[this very long discussion](https://github.com/conda-forge/pytorch-cpu-feedstock/issues/7)
+for details (and issues).
+
+Advantages of this alternative are:
+
+- Conda-forge already has a lot of packages, and there are established tools
+  and processes for adding and updating them. Which means it's less likely
+  for there to be issues with dependencies (e.g. packages with many or
+  unusual dependencies may not be accepted into the `pytorch` channel, while
+  `conda-forge` will be fine with them).
+- Users are likely already familiar with using the `conda-forge` channel.
+
+Disadvantages of this alternative are:
+
+- As of today, conda-forge doesn't have GPU hardware. Building is stil
+  possible using CUDA stubs, however testing cannot really be done inside CI,
+  only manually (which is a pain, especially when having to test multiple
+  hardware and OS platforms).
```

Comment:
To illustrate how this appoach of using CUDA stubs is used in practice, here is a complete list of conda-forge feedstocks that relay on CUDA stubs (meta.yaml contains `compiler("cuda")`):
[arrow-cpp](https://github.com/conda-forge/arrow-cpp-feedstock) - 3343 cuda downloads
[cupy](https://github.com/conda-forge/cupy-feedstock) - 2449 downloads
[faiss-split](https://github.com/conda-forge/faiss-split-feedstock) - 714 gpu downloads
[goofit-split](https://github.com/conda-forge/goofit-split-feedstock) - 942 downloads
[hoomd](https://github.com/conda-forge/hoomd-feedstock) - 2626 downloads
[implicit](https://github.com/conda-forge/implicit-feedstock) - 3725 downloads
[libtike-cufft](https://github.com/conda-forge/libtike-cufft-feedstock) - 796 downloads
[nccl](https://github.com/conda-forge/nccl-feedstock) - 15238 downloads
[nvidia-apex](https://github.com/conda-forge/nvidia-apex-feedstock) - 5584 downloads
[omniscidb](https://github.com/conda-forge/omniscidb-feedstock) - 678 cuda downloads
[openmm](https://github.com/conda-forge/openmm-feedstock) - 2884 downloads
[openmpi](https://github.com/conda-forge/openmpi-feedstock) - 32710 downloads
[pp-sketchlib](https://github.com/conda-forge/pp-sketchlib-feedstock) - 641 downloads
[prismatic_split](https://github.com/conda-forge/prismatic_split-feedstock) - 571 gpu downloads
[pytorch_cluster](https://github.com/conda-forge/pytorch_cluster-feedstock) - 532 downloads
[pytorch_scatter](https://github.com/conda-forge/pytorch_scatter-feedstock) - 575 downloads
[pytorch_sparse](https://github.com/conda-forge/pytorch_sparse-feedstock) - 528 downloads
[pytorch_spline_conv](https://github.com/conda-forge/pytorch_spline_conv-feedstock) - 514 downloads
[tomopy](https://github.com/conda-forge/tomopy-feedstock) - 2906 downloads
[ucx-split](https://github.com/conda-forge/ucx-split-feedstock) - 1253 gpu downloads

Here `downloads` is the number of downloads of the most popular version of the package. When the cuda or gpu specified, the package corresponds to the one built with CUDA stubs (and if not, the info is just not readily available).

Reply:
Thanks @pearu. A few thoughts:

- That list is a little long, I will summarize it and include a couple of the most important projects. 
- Do you have a script to generate this? Would be nice to reference, maybe as a gist (or include if it's short).
- The `pytorch_*` ones shouldn't (and don't, AFAICT) work, because there's no `pytorch` package in `conda-forge` so C++ code cannot link with the PyTorch C++ API . They have `build_locally.py` scripts, which may be what has been done in the past, just build locally and upload. A quick look at a recent `pytorch_sparse` build shows it picks up pytorch 1.4.0 from `defaults`.

Reply:
No script but the procedure was the following:
- clone https://github.com/conda-forge/feedstocks
- do `grep 'compiler(' feedstocks/*/meta.yaml | grep cuda` that will show all conda-forge feedstocks that use `compiler("cuda")`
- the list above was generated manually with some python scripting help

The list should be taken as it is. No guarantees on the relevance of the list items :)

Reply:
Adopted now, thanks